### ISSUE TITLE
[inductor] torch.compile silently drops an in-place mutation of an input made through F.dropout's identity return

### ISSUE THREAD

**Issue by laolvfan** [pytorch/pytorch#195422]
## 🐛 Describe the bug

When a function mutates its input in place *through a tensor returned by `F.dropout`*, `torch.compile` silently drops the mutation: the caller's tensor is left untouched, and any later use of that tensor inside the function computes on the stale value. No error, no warning, no graph break.

`F.dropout(x, p, training=False)` returns `x` itself, so `y += y` below is an in-place mutation of the input. Eager performs it; the compiled version does not.

Direct mutation of an input is handled correctly — `x += x`, `y = x; y += y`, and `y = x.view_as(x); y += y` all behave the same in both modes. Only the dropout-aliased form diverges.

## Reproduction

```python
import torch

def f(x):
    y = torch.nn.functional.dropout(x, 0.5, False)   # training=False -> y IS x
    y += y                                            # in-place mutation of the input
    return y

x = torch.ones(2, 2)
f(x)
print("eager    input after call:", x.flatten()[0].item())      # 2.0  (mutated)

x = torch.ones(2, 2)
torch.compile(f, backend="inductor")(x)
print("compiled input after call:", x.flatten()[0].item())      # 1.0  (mutation lost)
```

```
eager    input after call: 2.0
compiled input after call: 1.0
```

`p=0.0, training=True` behaves the same way, so this is not about the random mask:

```python
y = torch.nn.functional.dropout(x, 0.0, True)   # same result: mutation lost
```

## What works and what does not

All rows below are `y = <op>(x); y += y`, with `x = torch.ones(2, 2)`; the number is `x` after the call. Every listed op returns its input unchanged, so eager gives 2.0 in every row.

| form | eager | 2.11.0 compiled | 2.13.0 compiled |
|---|---|---|---|
| `F.dropout(x, 0.5, False)` | 2.0 | **1.0** | **1.0** |
| `F.dropout(x, 0.0, True)` | 2.0 | **1.0** | **1.0** |
| `F.alpha_dropout(x, 0.5, False)` | 2.0 | 2.0 | 2.0 |
| `nn.Identity()(x)` | 2.0 | 2.0 | 2.0 |
| `x.to(x.dtype)` | 2.0 | 2.0 | 2.0 |
| `x.contiguous()` | 2.0 | 2.0 | 2.0 |
| `y = x` (direct alias) | 2.0 | 2.0 | 2.0 |
| `x += x` (direct) | 2.0 | 2.0 | 2.0 |
| `x.view_as(x)` | 2.0 | 2.0 | 2.0 |

The scope is narrow: `F.alpha_dropout` is also the identity in eval mode and is handled correctly, as are `nn.Identity`, no-op `.to()`, `.contiguous()`, and plain aliases and views. Only `F.dropout` loses the alias between its output and its input, so the mutation is never attributed back to the graph input.

## Wrong results, not just a lost side effect

When the function keeps using the input after mutating it, the returned value is wrong too. Reduced from the program this was found in:

```python
import torch
import torch.nn as nn

class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = nn.Linear(1, 1)
        self.linear2 = nn.Linear(2, 2)

    def forward(self, x1):
        x2 = torch.nn.functional.dropout(x1, 0.0, True)
        x2 += x2
        x3 = torch.nn.functional.dropout(x1, 0.5, False)   # x3 IS x1
        x3 += x3                                            # x1 is now doubled
        a = self.linear1(x1)                                # eager sees the doubled x1
        b = self.linear2(torch.cat((x2, x3), dim=1).flatten())
        y = torch.add(a, b)
        y += y
        return y

torch.manual_seed(420)
m = M().eval()
x = torch.randn(1, 1, 1)

torch.manual_seed(420); print("eager   :", m(x).flatten().tolist())
torch.manual_seed(420)
x = torch.randn(1, 1, 1)
print("compiled:", torch.compile(m, backend="inductor")(x).flatten().tolist())
```

```
eager   : [8.649513244628906, -1.793665885925293]
compiled: [2.415100574493408, -3.086564540863037]
```

maxdiff 6.23 — silently wrong output, no exception.

## Versions

Reproduces on:

- PyTorch `2.11.0`
- PyTorch `2.13.0`

CPU, Linux x86_64, Python 3.11.

## Additional context

Checked nearby reports; I do not believe this is a duplicate:

- **#178586** (`make_graphed_callables` silently drops in-place mutations on user-provided input tensors) is about `torch.cuda.make_graphed_callables`, and its description states that `torch.compile` *does* detect input mutations and fall back safely. This report is a case where that detection misses: the mutation reaches the input through `F.dropout`'s identity return rather than directly.
- **#194747** (`aot_autograd: support input mutation on aliased inputs with different dtypes`) is about two graph inputs aliasing each other, not an intermediate aliasing an input.
- **#114142** (`Mutating graph inputs on torch.export fails`) is `torch.export` and closed.

Found by differential testing over StarCoder-generated programs; ten independently generated variants in the corpus hit the same divergence, and the reductions above come from them.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

**Comment by jansel** [comment]
IMO users should not be relying on dropout returning an alias to the input in some cases.

Are you seeing this in the wild or is this some sort of fuzzer issue?

**Comment by laolvfan** [comment]
Hi @jansel,

Yes, it's a fuzzer finding — differential testing over generated programs, not something I've hit in a real model. Fair to weigh the priority accordingly.

**Comment by jansel** [comment]
This no longer reproduces on main. It was fixed by #185335: the no-op branch of the `aten.dropout` decomposition (`train=False` or `p==0`) returned `input.clone()` where eager returns `input` itself, so functionalization attributed the `y += y` mutation to a fresh intermediate instead of the graph input and dropped the write-back. The fix first shipped in 2.14.0 — please retest there.

**Comment by laolvfan** [comment]
I retested the original behavior on Linux x86_64 CPU with PyTorch 2.14.0+cpu.

I used independent input tensors for eager and compiled execution and checked both the returned value and the caller-visible input mutation. Both identity branches now match eager:

- `F.dropout(x, 0.5, training=False)`
- `F.dropout(x, 0.0, training=True)`

In both cases, the returned tensor and the original input are correctly updated from all ones to all twos under `torch.compile`. I can confirm that the issue is fixed. Thanks for identifying the fixing PR.
